### DISCUSSION TITLE
How to add a transformer (BERT...) programmaticaly?

### DISCUSSION THREAD

**Discussion by asterbini** [explosion/spaCy#7954] [Help: Coding & Implementations] [answered]
I am trying to add a bert-large-cased transformer at the beginning of my pipeline programmaticaly.
Could I get a code example?

**Comment by adrianeboyd** [comment]
Hi, please see the docs here, in particular the examples in the right sidebar: https://spacy.io/api/transformer

**Comment by asterbini** [comment, marked as answer]
Yes, I tried this
```
import spacy
nlp = spacy.load('en_core_web_lg')
config = {
   "model": {
   "@architectures": "spacy-transformers.TransformerModel.v1",
   "name": "bert-large-cased",
   "tokenizer_config": {"use_fast": True}
   }
}
trf = nlp.add_pipe("transformer", config=config)
doc = nlp('Marcus loved Maria. Then she run away')
```
But I get an error 
```
... snip snip ...
/opt/miniconda3/envs/spacy3/lib/python3.9/site-packages/spacy_transformers/layers/transformer_model.py in forward(model, docs, is_train)
    119     tokenizer = model.attrs["tokenizer"]
    120     get_spans = model.attrs["get_spans"]
--> 121     transformer = model.layers[0]
    122 
    123     nested_spans = get_spans(docs)

IndexError: list index out of range
```

**Reply by adrianeboyd** [reply]
In spacy v3 most components also require an additional initialization step to load data and/or configure the statistical models, see: https://spacy.io/api/transformer#initialize. This is normally handled for you as part of `spacy train` if you're training a new model from a config.

As a note, adding a `transformer` component to `en_core_web_lg` is a fairly unusual thing to do, so if you want to provide more details about what you're trying to do, we can see if another approach might be better for your task?

**Reply by asterbini** [reply]
Thanks! 
By adding 
```
trf.initialize(lambda: iter([]), nlp=nlp)    # NOTICE the example of the doc sidebar is wrong
```
now I get my BERT running. 
Now from the tokens' contextual tensors/embedding I can disambiguate the Wordnet/Babelnet synsets by using ARES data (http://sensembert.org/)

**Reply by asterbini** [reply]
(but possibly the tokens will be unaligned respect to BERT ... I should dig more on that)

**Reply by adrianeboyd** [reply]
Thanks for the note, we will get that updated!

Check out the information about `TransformerData`: https://spacy.io/api/transformer#transformerdata

Be aware that the default config uses overlapping strided spans to deal with texts longer than the model max length, see some details in: #7032
